README.md
# Highscore
Highscore for mobilegame. In the website you can see top ten users and add a new user with points.
In the backend you can:
- Fetch all scores
- Fetch score based on id
- Fetch all scores with limit
- Fetch all scores with ascending or descending order
- Add a new score
- Delete a score by id
- Display high scores in html file format in a browser

The backend is secured with a hashed password.

# Author
Jonna Kyllönen, 
Matias Leppänen

# Screenshots
![Image of highscore](/highscore.png)


# Tech/framework used

- Python
- HTML
- CSS

# External modules:

- `flask`: a web framework for Python.
- `requests`: a third-party Python library for sending HTTP requests.
- `flask_bcrypt`: a third-party Python library for encrypting passwords using bcrypt hashing.
- `firebase_admin`: a third-party Python library for integrating with the Firebase platform.
- `Gunicorn`: Python HTTP server

# Installation and running:
- See the requirements.txt file or above which external modules are necessary to install before running the code and install them.
- Command for installing the external modules: pip install "module-name-here"

# clone highscore repository
- git clone "copy-the-repository-link-here" (More instructions from here: https://docs.github.com/en/repositories/creating-and-managing-repositories/cloning-a-repository)
- cd highscore

# start backend
- The file app.py cannot be run at the moment because of firebase. The file client.py uses the backend functions, so you can test them with the command:        python client.py 

# start frontend
https://scores-shxw.onrender.com/

# API implementation
API is deployed to cloud and can be accessed using following url:

https://scores-shxw.onrender.com/scores?pw=secret

# Screencast:
[![Screencast](https://img.youtube.com/vi/k4RMZif1Rdg/0.jpg)](https://www.youtube.com/watch?v=k4RMZif1Rdg)


release-notes.txt
RELEASE 1
- Fetching all scores
- Fetching score based on id
- Adding a new score
- Deleting a score by id
- Sort scores in asc or desc order

RELEASE 2
- Reading scores
- Save scores to the text file
- Limit scores
- Limit & sort work together
- HTML added

RELEASE 3
- Validations for username
- Password-function added for authentication
- HTML fixed to show only top ten users
- Tests for validation function

RELEASE 4 (Final release)
- HTML fixed to hide id from users
- HTML fixed to show top ten all the time
- Added Dropbox (and deleted it :D)
- Added Firebase
- Added client
- Pdoc
- Screencast made
- README updated



test_validation.py
import unittest
from validation import *


class TestValidation(unittest.TestCase):

    def test_validate_username(self):
        self.assertTrue(validate_username("Pimu"))
        self.assertFalse(validate_username("pimu"))
        self.assertFalse(validate_username("P imu"))  
        self.assertFalse(validate_username("P+mu")) 
        self.assertFalse(validate_username("P*imu"))
        self.assertFalse(validate_username(" Pimu")) 
        self.assertFalse(validate_username("Pimu ")) 
        self.assertFalse(validate_username("P1mu"))
        self.assertFalse(validate_username("Pimupimupimu"))
        self.assertFalse(validate_username("Kakka"))

client.py
import requests

while True:
    print("1) Add customer")
    print("2) Delete customer with id")
    print("3) Fetch all")
    print("4) Fetch scores of your choice")
    print("5) Fetch score based on ID")
    print("6) Quit")

    choice = int(input())

    url = 'http://scores-shxw.onrender.com/scores?pw=secret'

    if choice == 1:
        username = input("Enter the username: ")
        customer_points = int(input("Enter the customers points: "))
        myobj = {'name': username, 'points': customer_points}
        x = requests.post(url, json = myobj)
        if x.status_code == 200:
            print("Customer added succesfully")
        else:
            print("Your username is already in use or it is invalid, please choose another.")
        print(x.status_code)
    elif choice == 2:
        the_id=input("Give the customer id you want to delete:\n")
        d = requests.delete(f"http://scores-shxw.onrender.com/scores/{the_id}?pw=secret")
        if d.status_code == 204:
            print("Delete complete.")
        else:
            print(f"ID {the_id} not found")
        print(d.status_code)
    elif choice == 3:
        r = requests.get('http://scores-shxw.onrender.com/scores?pw=secret')
        print(r.text)
    elif choice == 4:
        sort_order = input("Sort in ascending or descending order? (asc/desc): ")
        limit = int(input("How many top scores do you want to fetch?: "))
        params = {"sort": sort_order, "limit": limit}
        r = requests.get(f"{url}?pw=secret", params=params)
        print(r.text)
    elif choice == 5:
        score_id = input("Enter the score id: ")
        r = requests.get(f"http://scores-shxw.onrender.com/scores/{score_id}?pw=secret")
        if r.status_code == 200:
            score = r.json()
            print(f"Score id: {score['id']}, Name: {score['name']}, Points: {score['points']}")
        else:
            print("Score not found.")
    elif choice == 6:
        print("Bye.")
        break
    else:
        print("Give a number between 1-6:")

repository.py
import os
from flask import jsonify, json
import firebase_admin
from firebase_admin import credentials
from firebase_admin import storage, firestore
import tempfile

# luetaan firebase ympäristömuuttuja
# tee render.comiin uusi muuttuja nimeltä firebase jonka
# sisältö on json tiedosto jonka saat firebaselta
json_str = os.environ.get('firebase')
"""
A JSON string containing Firebase credentials.
"""

# tallennetaan ympäristömuuttujan sisältö väliaikaiseen tiedostoon
with tempfile.NamedTemporaryFile(mode='w', delete=False) as f:
    f.write(json_str)
    temp_path = f.name
"""
A temporary file containing the Firebase credentials JSON string.
"""

# luetaan tiedostosta json filu
cred = credentials.Certificate(temp_path)
"""
Firebase credentials object.
"""

# tee render.comiin ympäristömuuttuja bucket, jonka sisältö
# esim: mydatabase-38cf0.appspot.com
firebase_admin.initialize_app(cred, {
    'storageBucket': os.environ.get('bucket')
})
bucket = storage.bucket()
"""
Firebase Storage bucket object.
"""

FILE = 'scores.txt'
"""
Name of the file that stores scores data.
"""

# JONNA read_scores
def read_scores():
    """
    Read scores data from Firebase Storage.

    Returns:
    A list of scores in JSON format.
    """
    try:
        blob = bucket.blob(FILE)
        scores_json = blob.download_as_string().decode('utf-8')
        return json.loads(scores_json)
    except:
        return []

# JONNA save_scores to the scores.txt
def save_to_scores(scores):
    """
    Save scores data to Firebase Storage.

    Args:
    - scores: A list of scores in JSON format.
    """
    with open('scores.txt', 'w') as f:
        blob = bucket.blob(FILE)
        blob.upload_from_string(json.dumps(scores), content_type='text/plain')


def main():
    """
    Print scores data.
    """
    print(read_scores())

if __name__ == "__main__":
    main()

# # KÄYTÄ NÄITÄ KOODEJA KUN TEET PDOCSIT: (KOMMENTOI KAIKKI YLLÄOLEVA POIS SIKSI AIKAA) TAI JOS HALUAT AJAA APP.PY !!!!!!
# import json

# FILE = 'scores.txt'
# """
# Name of the file that stores scores data.
# """

# # JONNA read_scores
# def read_scores():
#     """
#     Read scores data from the scores.txt file.

#     Returns:
#     A list of scores in JSON format.
#     """
#     try:
#         with open(FILE, 'r') as f:
#             scores_json = f.read()
#             return json.loads(scores_json)
#     except:
#         return []

# # JONNA save_scores to the scores.txt
# def save_to_scores(scores):
#     """
#     Save scores data to the scores.txt file.

#     Args:
#     - scores: A list of scores in JSON format.
#     """
#     with open(FILE, 'w') as f:
#         f.write(json.dumps(scores))


# def main():
#     """
#     Print scores data.
#     """
#     print(read_scores())

# if __name__ == "__main__":
#     main()

validation.py
import re
from typing import List
from repository import *

def validate_username(name):
    """
    Validates a username against a set of predefined rules.
    
    Args:
        name: A string containing the username to be validated.
    
    Returns:
        A boolean indicating whether the username is valid or not.
    """
    invalid_usernames: List[str] = ['Admin', 'Root', "Anal", "Anus", "Arse", "Ass", "Aallsack", "Aalls", "Bastard", "Bitch", "Biatch", "Bloody",
                                    "Blowjob", "Blow", "Bollock", "Bollok", "Boner", "Boob", "Bugger", "Bum", "Butt", "Buttplug", "Clitoris",
                                    "Cock", "Coon", "Crap", "Cunt", "Damn", "Dick", "Dildo", "Dyke", "Fag", "Feck", "Fellate", "Fellatio",
                                    "Felching", "Fuck", "Fudgepacker", "Flange", "Goddamn", "God damn", "Hell",
                                    "Homo", "Jerk", "Jizz", "Knobend", "Labia", "Lmao", "Lmfao", "Muff", "Nigger", "Nigga", "Omg",
                                    "Penis", "Piss", "Poop", "Prick", "Pube","Pussy", "Queer", "Scrotum", "Sex", "Shit", "Slut",
                                    "smegma", "Spunk", "Tit", "Tosser", "Turd", "Twat", "Vagina", "Wank", "Whore", "Wtf", "Huora", "Perse", "Paska",
                                    "Pillu", "Mulkku", "Vittu", "Kyrpä", "Neekeri", "Kusi", "Perkele", "Saatana", "Helvetti", "Lutka", "Kusipää",
                                    "Wittu", "Ripuli", "Kakka", "Seksi", "Pallihiki", "Palli", "Pylly", "Pimppi", "Nekru", "Sperma", "Kyrpa",
                                    "Kusipaa", "Tissi", "Tissit", "Mälli", "Idiootti", "Transu", "Trans", "Lesbo", "Hintti", "Vammanen",
                                    "Wammanen", "Kehari", "Boobs", ]
    pattern = r'^(?!.*(' + '|'.join(invalid_usernames) + '))[A-ZÄÖÅ][a-zA-ZäÄöÖåÅ]{2,9}$'
    if re.match(pattern, name):
        return True
    else:
        return False
    
def username_in_use(name):
    """
    Checks whether a given username is already in use.
    
    Args:
        name: A string containing the username to be checked.
    
    Returns:
        A boolean indicating whether the username is in use or not.
    """
    scores = read_scores()
    for score in scores:
        if score['name'] == name:
            return False
    return True


app.py
from flask import Flask, Response, jsonify, request, make_response, render_template, abort
import json
from operator import itemgetter
from repository import *
import requests
from validation import *
from flask_bcrypt import Bcrypt
import firebase_admin
from firebase_admin import credentials
from firebase_admin import storage
import tempfile


app = Flask(__name__)
bcrypt = Bcrypt(app)

# Replace 'secret' with the hashed version of the password
password_hash = bcrypt.generate_password_hash('secret').decode('utf-8')

# Password authentication decorator
def require_password(func):
    """
    Decorator function for password authentication.

    This function takes in a function as a parameter and returns a new function that wraps the original function with password authentication. If the correct password is not provided, an HTTPException with a 401 status code and an "Authentication required" message will be raised.

    Parameters:
    func (function): The function to be decorated with password authentication.

    Returns:
    function: A new function that wraps the original function with password authentication.

    Raises:
    HTTPException: If no password is provided or the provided password does not match the hashed password.

    """
    def wrapper(*args, **kwargs):
        """"
        A decorator function that wraps another function to provide authentication functionality. 

        Parameters
        ----------
        *args : tuple
            positional arguments that will be passed to the decorated function
        **kwargs : dict
            keyword arguments that will be passed to the decorated function

        Returns
        -------
        result : Any
        The result of the decorated function if the authentication is successful
        """
        # Get the password from the query parameters of the request
        pw = request.args.get("pw")
        #If no password is given OR the password doesn't match, the program aborts with error message
        if not pw or not bcrypt.check_password_hash(password_hash, pw):
            abort(401, "Authentication required")
        # If the password is correct, call the decorated function with the original arguments and return the result
        return func(*args, **kwargs)
    # Set the name of the wrapper function to be the same as the name of the endpoint function
    wrapper.__name__ = func.__name__
    return wrapper

@app.after_request
def after_request(response):
    """
    Adds headers to the response to enable Cross-Origin Resource Sharing (CORS).

    CORS is a mechanism that allows many resources (e.g., fonts, JavaScript, etc.) on a web page to be requested
    from another domain outside the domain the resource originated from. By default, web browsers block CORS requests.

    The headers added to the response enable CORS requests from any origin ('*') and allow the GET, POST, and DELETE
    methods to be used.

    Args:
        response (Flask response object): The response to add headers to.

    Returns:
        A Flask response object with headers added to enable CORS requests.
    """
    response.headers['Access-Control-Allow-Methods'] = 'GET, POST, DELETE'
    response.headers['Access-Control-Allow-Origin'] = '*'
    return response

# JONNA & MATIAS "Fetching all scores":
@app.route("/scores")
@require_password
def get_scores():
    """
    Retrieve and sort a list of scores from a Cloud Storage bucket, based on the provided query parameters.

    Args:
    -None
    
    Returns:
    - A JSON response containing a sorted and limited list of score objects, with an HTTP status code of 200 (OK).

    Raises:
    - None
    
    HTTP Route:
    - GET '/scores'
    
    HTTP Query Parameters:
    - sort: string value indicating how to sort the scores list. Possible values are "asc" (for ascending order) and "desc" (for descending order).
    - limit: integer value indicating the maximum number of scores to return.

Example Usage:
    To retrieve the top 10 scores in descending order:
    GET '/scores?sort=desc&limit=10'
"""
    
    # Retrieve the 'scores.txt' blob from the default bucket associated with the Cloud Storage client (Firebase)
    blob = bucket.blob('scores.txt')
    # Download the contents of the blob as a byte string, decode it as UTF-8 text, and parse it as JSON to create a list of dictionaries
    scores = blob.download_as_string().decode('utf-8')
    scores = json.loads(scores)
    
    # Copy the scores list to avoid modifying the original list.
    sorted_scores = scores
    # Get the 'sort' query parameter from the request, which specifies how to sort the scores.
    sort = request.args.get("sort")
    # If the 'sort' parameter is set to 'asc', sort the scores list in ascending order.
    if sort == "asc":
        sorted_scores = sorted(sorted_scores, key=lambda s: int(s["points"]))
    # If the 'sort' parameter is set to 'desc', sort the scores list in descending order.
    elif sort == "desc":
        sorted_scores = sorted(sorted_scores, key=lambda s: int(s["points"]), reverse=True)
    # Get the 'limit' query parameter from the request, which specifies the maximum number of scores to return.
    limit = request.args.get("limit")
    # If the 'limit' parameter is set, slice the sorted scores list to return only the first 'limit' scores.
    if limit:
        sorted_scores = sorted_scores[:int(limit)]
    # Return the sorted and limited scores list as a JSON response.
    return jsonify(sorted_scores)

# JONNA "Fetching score based on id": 
@app.route('/scores/<int:the_id>')
@require_password
def get_scores_id(the_id):
    """
    Returns the score object with the given id as a JSON response.

    Args:
    - the_id: integer value representing the id of the desired score object

    Returns:
    - A JSON response containing the score object with the given id and an HTTP status code of 200 (OK) if a matching score object is found.
    - An empty response and an HTTP status code of 404 (Not Found) if no matching score object is found.
    """
    scores = read_scores()
    for score in scores:
        # Check if the 'id' of the current score matches the given 'the_id'
        if score["id"] == the_id:
            # If a match is found, return the score object as a JSON response with an HTTP status code 
            return jsonify(score), 200
    return make_response ("", 404)

# MATIAS Created a function for getting the 10th score so that in the game the app will compare the players points to the 10th score.
@app.route('/scores/last_score')
def get_last_score():
    """
    Returns the score object with the 10th 'points' value. This was made for Unity, so that it would check if the player gets to the top 10 list.

    Returns:
    - A JSON response containing the score object with the highest 'points' value, with an HTTP status code of 200 (OK).
    """
    scores = read_scores()
    # Sort the scores in descending order by 'points'
    scores=sorted(scores, key=lambda k: int(k['points']), reverse=True)
    last_score=scores[9]
    return jsonify(last_score), 200

# # JONNA "Adding a new score" (Backend) 
@app.route('/scores', methods=['POST'])
@require_password
def add_score():
    """
    Adds a new score to the list of scores and generates ID.

    Args:
        None

    Returns:
        - A response with an HTTP status code of 200 (OK) if the score is added successfully.
        - A response with an HTTP status code of 400 (Bad Request) if the JSON data is invalid.
    """
    # # load given JSON data and turn it into dictionary
    score = request.get_json()
    if not score:
         return make_response("Invalid JSON data", 400)
    
         # validate username
    username = score.get('name')
    if not validate_username(username):
         return make_response("Invalid username", 400)
    if not username_in_use(username):
         return make_response("Username already in use", 400)
    
    scores = read_scores()
     # generate new score ID
    if scores:
        score_id = scores[-1]['id'] + 1
    else:
        score_id = 1
     # add new score with generated ID
    score['id'] = score_id
    scores.append(score)
    # save updated scores
    save_to_scores(scores)
    # return success response
    return jsonify(scores)


# JONNA "Deleting a score by id":
@app.route('/scores/<int:the_id>', methods=['DELETE'])
@require_password
def delete_customer(the_id):
    """
    Deletes the score object with the given id.

    Args:
    - the_id (int): the id of the score object to be deleted.

    Returns:
    - A tuple containing an empty response and an HTTP status code of 204 (No Content) if a matching score object is found and deleted.
    - A tuple containing an empty response and an HTTP status code of 404 (Not Found) if no matching score object is found.

    Raises:
    - ValueError: if the_id parameter is not an integer.

    """
    scores = read_scores()
    # Reads the scores from the database

    index_to_be_deleted = -1
    # Initializes the index of the score object to be deleted to -1

    for i in range(0, len(scores)):
        if(scores[i]["id"] == the_id):
            index_to_be_deleted = i
    # Loops through the scores list to find the index of the score object with the given id.       
    
    if(index_to_be_deleted != -1):
        scores.pop(index_to_be_deleted)
          # Removes the score object at the identified index from the scores list.

        save_to_scores(scores)
         # Saves the updated scores to the database.

        return make_response("", 204)
        # Returns a response with status code 204 (No Content) to indicate that the score object was successfully deleted.

    else:
        return make_response("", 404)
        # Returns a response with status code 404 (Not Found) to indicate that no matching score object was found.

# JONNA add new score (Frontend)
# @app.route('/scores', methods=['POST'])
def add_score_to_database(score):
    """
    Add a new score to the database.

    Args:
        score (dict): The new score to add to the database.

    Returns:
        bool: True if the score was successfully added to the database, False otherwise.
    """
    scores = read_scores()
    # generate new score ID
    if scores:
        score_id = scores[-1]['id'] + 1
    else:
        score_id = 1
    # add new score with generated ID
    score['id'] = score_id
    scores.append(score)
    # save updated scores
    save_to_scores(scores)
    return True

@app.route('/', methods = ['POST', 'GET'])
def index():
    """
    The index function that handles both GET and POST requests to the root route. When a GET request is made,
    the function reads the data from the score-file and renders the index.html template with a form.
    When a POST request is made, the function reads the form data, validates the name and saves the data to the database.
    It then reads the updated data from the score-file and renders the index.html template with the updated data.

    Returns:
        str: the rendered HTML template as a string.
    """ 

    if request.method == 'POST':
        name = request.form.get('name')
        points = request.form.get('points')
        score = {'name': name, 'points': points}
        # Get the values of 'name' and 'points' from the form data and create a dictionary 'score' containing these values.
         
        if not validate_username(name):
            scores = read_scores()
            sorted_scores = sorted(scores, key=lambda x: int(x['points']), reverse=True)[:10]
            return render_template('scores.html', error='Username may contain up to ten letters without special characters and the first letter must be capitalized', scores=sorted_scores)
        # If the name is not valid (according to the 'validate_username' function), return 'scores.html' template with an error message.
        if not username_in_use(name):
            scores = read_scores()
            sorted_scores = sorted(scores, key=lambda x: int(x['points']), reverse=True)[:10]
            return render_template('scores.html', error='Username is already in use.', scores=sorted_scores)


        score = {'name': name, 'points': points}
        # Create a new dictionary 'score' containing the name and points values.
       
        if add_score_to_database(score):
            scores = read_scores()
            sorted_scores = sorted(scores, key=lambda x: int(x['points']), reverse=True)[:10]
            return render_template('scores.html', scores=sorted_scores)
        else:
            return "Failed to save score", 500
        # If the score was added to the database successfully, read the updated scores from the database
        # and render the 'scores.html' template with the updated scores. Otherwise, return an error message.
    else:
        scores = read_scores()
        scores = sorted(scores, key=lambda k: int(k['points']), reverse=True)[:10]
        return render_template('scores.html', scores=scores)
        # If the request method is not POST, read the scores from the database and show 'scores.html' template with the scores.

if __name__ == "__main__":
    app.run()

